Storage/storage.h

```diff
@@ -73,22 +73,47 @@ namespace storage {
+    CURRENT_CONSTRUCTOR(alias##Updated)(const base& value) : data(value) {}                           \
+    using T_KEY = current::decay<decltype(::current::storage::sfinae::GetKey(std::declval<base>()))>; \
+    T_KEY key() const { return ::current::storage::sfinae::GetKey(data); }                            \
+    void set_key(const T_KEY& key) { ::current::storage::sfinae::SetKey(data, key); }                 \
+  };                                                                                                  \
+  CURRENT_STRUCT(alias##Deleted) {                                                                    \
+    CURRENT_FIELD(key, ::current::storage::sfinae::ENTRY_KEY_TYPE<base>);                             \
+    CURRENT_DEFAULT_CONSTRUCTOR(alias##Deleted) {}                                                    \
+    CURRENT_CONSTRUCTOR(alias##Deleted)(const base& value)                                            \
+        : key(::current::storage::sfinae::GetKey(value)) {}                                           \
+  };                                                                                                  \
+  struct alias {                                                                                      \
+    using T_ENTRY = base;                                                                             \
+    using T_UPDATE_EVENT = alias##Updated;                                                            \
+    using T_DELETE_EVENT = alias##Deleted;                                                            \
+    using T_PERSISTED_EVENT_1 = alias##Updated;                                                       \
+    using T_PERSISTED_EVENT_2 = alias##Deleted;                                                       \
   }
+
+#define CURRENT_STRUCT_TAG_Vector(base, alias)                                    \
+  CURRENT_STRUCT(alias##PushedElement) {                                          \
+    CURRENT_FIELD(data, base);                                                    \
+    CURRENT_DEFAULT_CONSTRUCTOR(alias##PushedElement) {}                          \
+    CURRENT_CONSTRUCTOR(alias##PushedElement)(const base& value) : data(value) {} \
+  };                                                                              \
+  CURRENT_STRUCT(alias##PoppedElement) {};                                        \
+  struct alias {                                                                  \
+    using T_ENTRY = base;                                                         \
+    using T_PUSHED_BACK_EVENT = alias##PushedElement;                             \
+    using T_POPPED_BACK_EVENT = alias##PoppedElement;                             \
+    using T_PERSISTED_EVENT_1 = alias##PushedElement;                             \
+    using T_PERSISTED_EVENT_2 = alias##PoppedElement;                             \
+  }
+
+// TODO(dkorolev) || TODO(mzhurovich): Move to `TypeSystem/` and test.
```

Comment:
Honestly speaking, I'm strongly against adding dependencies to Current components if we may not to.
If I got it right, you propose to move some Storage-related code to the `TypeSystem`. Why not just left it in the Storage?